### PR TITLE
add initial modal bi-encoder app

### PR DESCRIPTION
**Pull request by justinliang1020**
# Add initial modal bi-encoder app

## Features

FastAPI app
* deployable on Modal
* endpoints include `/encode` and `/cosine_similarity`
* will include `finetune` endpoint in next PR

README
* Summary
* Installation instructions
* Usage instructions and example calls for both local/modal deployment

## Future features

Finetune endpoint
* finetune the sentence transformer model given input data
Will be based on code here: https://www.sbert.net/docs/training/overview.html#loss-functions
```python
from sentence_transformers import SentenceTransformer, InputExample, losses
from torch.utils.data import DataLoader

#Define the model. Either from scratch of by loading a pre-trained model
model = SentenceTransformer('distilbert-base-nli-mean-tokens')

#Define your train examples. You need more than just two examples...
train_examples = [InputExample(texts=['My first sentence', 'My second sentence'], label=0.8),
    InputExample(texts=['Another pair', 'Unrelated sentence'], label=0.3)]

#Define your train dataset, the dataloader and the train loss
train_dataloader = DataLoader(train_examples, shuffle=True, batch_size=16)
train_loss = losses.CosineSimilarityLoss(model)

#Tune the model
model.fit(train_objectives=[(train_dataloader, train_loss)], epochs=1, warmup_steps=100)
``` 
* will create a new `src/datasets.py` file to handle importing data
* modify existing `/encode` and `/cosine_similarity` endpoints to be able to use the new finetuned model
* figure out how to make the finetuned model downloadable locally and on modal

### CODE REVIEW

File: applications/modal-bi-encoder/app.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    CosineSimilarityInput,
+    CosineSimilarityOutput,
+)
+
+app = FastAPI(
+    title="Sentence Transformer: Bi-Encoder",
+)
+
+
+@app.post("/encode", response_model=EncodeOutput)
```
Comment (jxnl) [inline applications/modal-bi-encoder/app.py]: this should look like the openai embedding spec https://platform.openai.com/docs/api-reference/chat/create

File: applications/modal-bi-encoder/app.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+def encode_request(input: EncodeInput):
+    return encode(input)
+
+
+@app.post("/cosine_similarity", response_model=CosineSimilarityOutput)
+def cosine_similarity_request(input: CosineSimilarityInput):
```
Comment (jxnl) [inline applications/modal-bi-encoder/app.py:25]: this should look like the cohere api, check out my `sentence transformer` examples

File: applications/modal-bi-encoder/app.py (hunk 1 above)
Comment (justinliang1020) [reply applications/modal-bi-encoder/app.py]: Changed the `encode` endpoint to be more like the openai embeddings endpoint https://platform.openai.com/docs/api-reference/embeddings/create
* renamed to `embeddings`
* changed types to model openai output, inspired from this file: <https://github.com/jxnl/fastllm/blob/main/applications/embeddings-gpt4all/src.py>

OpenAPI docs examples with new endpoints: 
image
image

File: applications/modal-bi-encoder/app.py (hunk 2 above)
Comment (justinliang1020) [reply applications/modal-bi-encoder/app.py:25]: For this, there isn't a correlating API endpoint for `CosineSimilarity` in the Cohere (or OpenAI) API.

There is a [reranking function](https://docs.cohere.com/reference/rerank-1) in Cohere that is in the `sentence transformers` example, however since this current directory, `modal-bi-encoder`, is only for bi-encoders, I don't believe it should implement any reranking function as that's mainly for cross encoders. <https://www.sbert.net/examples/applications/retrieve_rerank/README.html> Thus, the Cohere-like reranking endpoint will instead be implemented in a future PR instead of this one, where the new `modal-cross-encoder` directory will be made that deals with fine-tuning a cross encoder.

So, I believe it's ok to leave the `cosine_similarity` function as is since there isn't a Cohere correlating function to match the API specs to.

File: applications/modal-bi-encoder/app.py (hunk 1 above)
Comment (justinliang1020) [reply applications/modal-bi-encoder/app.py]: I also updated the README to reflect these changes

File: applications/modal-bi-encoder/app.py (hunk 2 above)
Comment (jxnl) [reply applications/modal-bi-encoder/app.py:25]: sounds good to me

File: applications/modal-bi-encoder/src/main.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+def get_model(model_name=MODEL_NAME) -> SentenceTransformer:
+    return SentenceTransformer(model_name)
+
+
+def embeddings(req: EmbeddingInput) -> EmbeddingOutput:
+    """Embed texts using the given sentence embedding model
+    Source: https://www.sbert.net/docs/quickstart.html"""
+    model = get_model(req.model)
+    sentences = req.input if isinstance(req.input, List) else [req.input]
+    embeddings = model.encode(sentences, convert_to_tensor=True)
```
Comment (jxnl) [inline applications/modal-bi-encoder/src/main.py:60]: the encode function shuold be a stub.function like this: 

https://github.com/modal-labs/modal-examples/blob/main/06_gpu_and_ml/embeddings/text_embeddings_inference.py#L71-L100

**Comment by justinliang1020** [comment]
outdated PR, see here for the updated one: https://github.com/567-labs/fastllm/pull/28
